src/main/java/bio/terra/service/dataset/DatasetBucketDao.java

```diff
@@ -0,0 +1,104 @@
+package bio.terra.service.dataset;
+
+import bio.terra.common.exception.RetryQueryException;
+import bio.terra.service.snapshot.exception.CorruptMetadataException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.dao.DataAccessException;
+import org.springframework.jdbc.core.namedparam.MapSqlParameterSource;
+import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate;
+import org.springframework.stereotype.Repository;
+import org.springframework.transaction.annotation.Isolation;
+import org.springframework.transaction.annotation.Propagation;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.UUID;
+
+import static bio.terra.common.DaoUtils.retryQuery;
+
+@Repository
+public class DatasetBucketDao {
+    private static final Logger logger = LoggerFactory.getLogger(DatasetBucketDao.class);
+
+    // Note we start from 1, since we are recording an ingest creating the link. If that ingest fails
+    // it will decrement to zero.
```

Comment:
The ingest counter is a mechanism to avoid locking the dataset-bucket row. The only important values are 0 or not 0. A row with a counter of 0 is equivalent to no row; that is, no files in the dataset are using the bucket. A value greater than 0 means the dataset is using the bucket.

An ingest either creates the row with a value of 1 or it increments the value. The undo step always decrements the value.
That way, if there are parallel ingests and one fails, it doesn't have to try to delete the row. It just decrements.

The alternative would be what we do in other objects like datasets. Add an id column and a flightid column. The creator mints an id and does a create_and_lock. At the end of the successful flight, the creator unlocks. Another flight will wait (for the whole ingest to complete) until the lock is unlocked (or the row is deleted).

Reply:
I'll add this as documentation in the code.